File: src/utils/crypto.ts
// crypto.ts
import CryptoJS from "crypto-js";

// 定义配置接口
interface CryptoConfig {
  readonly SECRET_KEY: string;
  readonly IV: string;
}

//AES-256-CBC

// 定义加密选项接口
interface CryptoOptions {
  iv: CryptoJS.lib.WordArray;
  mode: typeof CryptoJS.mode.CBC;
  padding: typeof CryptoJS.pad.Pkcs7;
}

// 定义加密服务类
export class CryptoService {
  private readonly key: CryptoJS.lib.WordArray;
  private readonly iv: CryptoJS.lib.WordArray;

  constructor(config: CryptoConfig) {
    if (config.SECRET_KEY.length !== 32) {
      throw new Error("SECRET_KEY must be 32 characters long");
    }
    if (config.IV.length !== 16) {
      throw new Error("IV must be 16 characters long");
    }

    this.key = CryptoJS.enc.Utf8.parse(config.SECRET_KEY);
    this.iv = CryptoJS.enc.Utf8.parse(config.IV);
  }

  private get encryptOptions(): CryptoOptions {
    return {
      iv: this.iv,
      mode: CryptoJS.mode.CBC,
      padding: CryptoJS.pad.Pkcs7
    };
  }

  /**
   * 加密文本
   * @param text 要加密的文本
   * @returns 加密后的字符串
   * @throws Error 如果加密失败
   */
  public encrypt(text: string): string {
    try {
      if (!text) {
        throw new Error("Text to encrypt cannot be empty");
      }

      const encrypted = CryptoJS.AES.encrypt(
        text,
        this.key,
        this.encryptOptions
      );

      return encrypted.toString();
    } catch (error) {
      throw new Error(`Encryption failed: ${(error as Error).message}`);
    }
  }

  /**
   * 解密文本
   * @param encryptedText 加密的文本
   * @returns 解密后的原文
   * @throws Error 如果解密失败
   */
  public decrypt(encryptedText: string): string {
    try {
      if (!encryptedText) {
        throw new Error("Text to decrypt cannot be empty");
      }

      const decrypted = CryptoJS.AES.decrypt(
        encryptedText,
        this.key,
        this.encryptOptions
      );

      const result = decrypted.toString(CryptoJS.enc.Utf8);

      if (!result) {
        throw new Error("Decryption resulted in empty string");
      }

      return result;
    } catch (error) {
      throw new Error(`Decryption failed: ${(error as Error).message}`);
    }
  }
}
// 创建单例实例
const cryptoConfig: CryptoConfig = {
  SECRET_KEY: "mK9bP2vN8xL5tR7hJ4fD1cA3gE6iQ0wS", // 32字符密钥
  IV: "uY5nM2kX7pJ9vB4c" // 16字符IV
};

export const cryptoService = new CryptoService(cryptoConfig);

File: src-tauri/src/models/crypto.rs
use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use serde::{Deserialize, Serialize};
use thiserror::Error;

type Aes256CbcDec = cbc::Decryptor<aes::Aes256>;

#[derive(Debug, Error)]
pub enum DecryptError {
    #[error("Failed to decode base64: {0}")]
    Base64DecodeError(#[from] base64::DecodeError),

    #[error("Decryption failed: {0}")]
    DecryptionError(String),

    #[error("Invalid input: {0}")]
    InvalidInput(String),

    #[error("UTF-8 conversion failed: {0}")]
    Utf8Error(#[from] std::string::FromUtf8Error),
}

pub struct CryptoService {
    key: [u8; 32],
    iv: [u8; 16],
}

impl CryptoService {
    pub fn new(secret_key: &str, iv: &str) -> Result<Self, DecryptError> {
        // 验证密钥长度
        if secret_key.len() != 32 {
            return Err(DecryptError::InvalidInput(
                "SECRET_KEY must be 32 characters long".to_string(),
            ));
        }

        // 验证IV长度
        if iv.len() != 16 {
            return Err(DecryptError::InvalidInput(
                "IV must be 16 characters long".to_string(),
            ));
        }

        let key = secret_key.as_bytes().try_into().map_err(|_| {
            DecryptError::InvalidInput("Failed to convert key to bytes".to_string())
        })?;

        let iv = iv
            .as_bytes()
            .try_into()
            .map_err(|_| DecryptError::InvalidInput("Failed to convert IV to bytes".to_string()))?;

        Ok(Self { key, iv })
    }

    pub fn decrypt(&self, encrypted_text: &str) -> Result<String, DecryptError> {
        // 验证输入
        if encrypted_text.is_empty() {
            return Err(DecryptError::InvalidInput(
                "Encrypted text cannot be empty".to_string(),
            ));
        }

        // Base64解码
        let encrypted_data = BASE64.decode(encrypted_text)?;

        // 创建解密器
        let cipher = Aes256CbcDec::new(&self.key.into(), &self.iv.into());

        // 解密
        let mut buf = encrypted_data.to_vec();
        let decrypted_data = cipher
            .decrypt_padded_mut::<Pkcs7>(&mut buf)
            .map_err(|e| DecryptError::DecryptionError(e.to_string()))?;

        // 转换为UTF-8字符串
        let decrypted_text = String::from_utf8(decrypted_data.to_vec())?;

        Ok(decrypted_text)
    }
}

// 为了方便在命令中使用，可以创建一个包含配置的结构体
#[derive(Debug, Serialize, Deserialize)]
pub struct CryptoConfig {
    secret_key: String,
    iv: String,
}

// 创建全局实例
lazy_static::lazy_static! {
    pub static ref CRYPTO_SERVICE: CryptoService = CryptoService::new(
        "mK9bP2vN8xL5tR7hJ4fD1cA3gE6iQ0wS",  // 32字符密钥
        "uY5nM2kX7pJ9vB4c"   // 16字符IV
    ).expect("Failed to initialize crypto service");
}
